piptools/scripts/compile.py

```diff
@@ -387,16 +406,7 @@ def cli(
         ireq for key, ireq in upgrade_install_reqs.items() if key in allowed_upgrades
     )
 
-    # Filter out pip environment markers which do not match (PEP496)
-    constraints = [
-        req
-        for req in constraints
-        if req.markers is None
-        # We explicitly set extra=None to filter out optional requirements
-        # since evaluating an extra marker with no environment raises UndefinedEnvironmentName
-        # (see https://packaging.pypa.io/en/latest/markers.html#usage)
-        or req.markers.evaluate({"extra": None})
-    ]
+    constraints = [req for req in constraints if req_check_markers(req, extras=extras)]
```

Comment:
[InstallRequirement.match_markers(extras)](https://github.com/pypa/pip/blob/7ec0fa5142466f8402bb866a1dcfb8ca6ffdc66b/src/pip/_internal/req/req_install.py#L268) could be used instead of `req_check_markers()`. Looks like it works the same way under the hoods.
```suggestion
    constraints = [req for req in constraints if req.match_markers(extras)]
```